File: src/main/java/org/simbrain/plot/timeseries/TimeSeriesPlotComponent.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.plot.timeseries;

import com.thoughtworks.xstream.XStream;
import org.simbrain.plot.XYSeriesConverter;
import org.simbrain.util.DoubleArrayConverter;
import org.simbrain.util.XStreamUtils;
import org.simbrain.workspace.AttributeContainer;
import org.simbrain.workspace.Workspace;
import org.simbrain.workspace.WorkspaceComponent;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

/**
 * Represents time series data.
 */
public class TimeSeriesPlotComponent extends WorkspaceComponent {

    /**
     * The data model.
     */
    private final TimeSeriesModel model;

    /**
     * Create new time series plot component.
     *
     * @param name name
     */
    public TimeSeriesPlotComponent(String name) {
        super(name);
        model = new TimeSeriesModel(() -> getWorkspace().getTime());
    }

    /**
     * Creates a new time series component from a specified model. Used in
     * deserializing.
     *
     * @param name  chart name
     * @param model chart model
     */
    public TimeSeriesPlotComponent(String name, TimeSeriesModel model) {
        super(name);
        this.model = model;
        model.setTimeSupplier(() -> getWorkspace().getTime());
    }

    @Override
    public void setWorkspace(Workspace workspace) {
        // Workspace object is not available in the constructor.
        super.setWorkspace(workspace);

        getWorkspace().getCouplingManager().getEvents().getCouplingAdded().on(c -> {
            // A new array coupling is being added to this time series
            if (c.getConsumer().getBaseObject() == model) {

                // Initialize series with provided names, e.g neuron labels
                model.initializeArrayMode(c.getProducer().getLabelArray());
            }
        });

        model.getEvents().getChangeArrayMode().on(() -> {
            // Array mode has been changed
            if (model.isArrayMode()) {
                // Changed from scalar to array mode
                // No action
            } else {
                // Changed from array to scalar mode
                fireAttributeContainerRemoved(model);
            }
        });

        // A new scalar time series has been added
        model.getEvents().getScalarTimeSeriesAdded().on(this::fireAttributeContainerAdded);

        // A scalar time series has been removed
        model.getEvents().getScalarTimeSeriesRemoved().on(this::fireAttributeContainerRemoved);
    }

    @Override
    public List<AttributeContainer> getAttributeContainers() {
        List<AttributeContainer> containers = new ArrayList<>();
        if (model.isArrayMode()) {
            containers.add(model);
        } else {
            containers.addAll(model.getTimeSeriesList());
        }
        return containers;
    }

    public TimeSeriesModel getModel() {
        return model;
    }

    /**
     * Opens a saved time series plot.
     *
     * @param input  stream
     * @param name   name of file
     * @param format format
     * @return bar chart component to be opened
     */
    public static TimeSeriesPlotComponent open(final InputStream input, final String name, final String format) {
        TimeSeriesModel dataModel = (TimeSeriesModel) getTimeSeriesXStream().fromXML(input);
        return new TimeSeriesPlotComponent(name, dataModel);
    }

    @Override
    public void save(final OutputStream output, final String format) {
        getTimeSeriesXStream().toXML(model, output);
    }

    @Override
    public boolean hasChangedSinceLastSave() {
        return false;
    }

    @Override
    public String getXML() {
        return getTimeSeriesXStream().toXML(model);
    }

    public static XStream getTimeSeriesXStream() {
        var xstream = XStreamUtils.getSimbrainXStream();
        xstream.registerConverter(new DoubleArrayConverter());
        xstream.registerConverter(new XYSeriesConverter());
        return xstream;
    }


}

File: src/main/kotlin/org/simbrain/plot/timeseries/TimeSeriesPlotComponent.kt
package org.simbrain.plot.timeseries

import com.thoughtworks.xstream.XStream
import kotlinx.coroutines.Dispatchers
import org.simbrain.plot.XYSeriesConverter
import org.simbrain.plot.timeseries.TimeSeriesModel.TimeSeries
import org.simbrain.util.DoubleArrayConverter
import org.simbrain.util.getSimbrainXStream
import org.simbrain.workspace.*
import java.io.InputStream
import java.io.OutputStream

class TimeSeriesPlotComponent @JvmOverloads constructor(name: String, val model: TimeSeriesModel = TimeSeriesModel()) : WorkspaceComponent(name) {

    /** Whether scalar couplings automatically replace series names with their producers' names. */
    var useScalarCouplingNames = true

    override var workspace: Workspace
        get() = super.workspace
        set(workspace) {
            // Workspace object is not available in the constructor.
            super.workspace = workspace

            onCoupledProducer { consumer, producer ->
                if (consumer === model) {
                    model.syncTimeSeries(producer.displayComponents)
                } else if (consumer is TimeSeries && useScalarCouplingNames) {
                    applyScalarNames()
                }
            }

            model.events.timeSeriesAdded.on(Dispatchers.Default) { addedContainer: TimeSeries ->
                this.fireAttributeContainerAdded(addedContainer)
            }

            model.events.timeSeriesRemoved.on(Dispatchers.Default) { removedContainer: TimeSeries ->
                this.fireAttributeContainerRemoved(removedContainer)
            }
        }

    /**
     * Name the series that are themselves targets of scalar couplings, as produced by plotting a selection of
     * neurons. The producing attribute's name is enough on its own when every such series comes from the same
     * kind of attribute, e.g. all activations; the method name is appended only when it is what distinguishes
     * them, e.g. one neuron's activation plotted alongside its bias. They are named as a set because adding a
     * coupling can change which of those two forms applies to the series already present.
     */
    private fun applyScalarNames() {
        val scalarCouplings = couplingManager.couplings
            .filter { coupling -> model.timeSeriesList.any { it === coupling.consumer.baseObject } }
        if (scalarCouplings.isEmpty()) return
        val methodDistinguishes = scalarCouplings.map { it.producer.simpleMethodName }.toSet().size > 1
        val names = scalarCouplings
            .mapIndexed { index, coupling ->
                val producer = coupling.producer
                AttributeComponent(
                    "$index",
                    if (methodDistinguishes) producer.simpleDescription else producer.containerDisplayName
                )
            }
            .disambiguateNames()
            .map { it.name }
        model.renameEachTimeSeries(scalarCouplings.map { it.consumer.baseObject as TimeSeries }.zip(names))
    }

    override val attributeContainers: List<AttributeContainer>
        get() {
            val containers: MutableList<AttributeContainer> = ArrayList()
            containers.add(model)
            containers.addAll(model.timeSeriesList)
            return containers
        }

    fun addTimeSeries(name: String) = model.addTimeSeries(name)

    override fun save(output: OutputStream, format: String?) {
        timeSeriesXStream.toXML(model, output)
    }

    override fun hasChangedSinceLastSave(): Boolean {
        return false
    }

    override val xml: String
        get() = timeSeriesXStream.toXML(model)

    init {
        model.timeSupplier = { workspace.time }
    }

    companion object {
        /**
         * Opens a saved time series plot.
         *
         * @param input  stream
         * @param name   name of file
         * @param format format
         * @return bar chart component to be opened
         */
        fun open(input: InputStream, name: String, format: String?): TimeSeriesPlotComponent {
            val dataModel = timeSeriesXStream.fromXML(input) as TimeSeriesModel
            return TimeSeriesPlotComponent(name, dataModel)
        }

        val timeSeriesXStream: XStream
            get() {
                val xstream = getSimbrainXStream()
                xstream.registerConverter(DoubleArrayConverter())
                xstream.registerConverter(XYSeriesConverter())
                return xstream
            }
    }
}
